lib/core/jmanagedbackends.c

```diff
@@ -0,0 +1,832 @@
+		wrapper->scope = (JManagedBackendScope){
+			.mem = wrapper,
+			.lock_id = entry->lock_id,
+			.tier = entry->backend_id,
+			.stack = this,
+			.namespace = namespace,
+			.path = path,
+		};
+		*scope = &wrapper->scope;
+	}
+	else
+	{
+		EXE(j_backend_managed_object_create(this, namespace, path, backend, scope),
+		    "failed to create new entry");
+	}
+	ret = TRUE;
+end:
+	return ret;
+}
+
+gboolean
+j_backend_managed_object_create(JManagedBackends* this,
+				const gchar* namespace,
+				const gchar* path,
+				JBackend** backend,
+				JManagedBackendScope** scope)
+{
+	gboolean ret = FALSE;
+	struct JBackendWrapper* wrapper;
+	struct KVEntry entry;
+	guint lock_id;
+	guint tier;
+	RWSpinLock newLock;
+
+	// allocate new spinlock with read lock
+	newLock = (RWSpinLock){ .generation = 0, .read_access_count = 1, .write_access = 0 };
+	g_array_append_val(this->rw_spin_locks, newLock);
+	lock_id = g_atomic_int_add(&this->array_length, 1);
+
+	// ask policy for tier
+	EXE(this->policy->process_create(this->policy->data, namespace, path, lock_id, &tier),
+	    "failed to match storage tier for new object!");
+	*backend = g_memdup2(this->object_backend[tier], sizeof(struct JBackendWrapper));
+	wrapper = (void*)*backend;
+	if (this->log.filename)
+	{
+		guint64 i = g_atomic_int_add(&this->log.length, 1);
+		if (i < MAX_LOG_LENGTH)
+		{
+			this->log.entries[i].access = 'c';
+			this->log.entries[i].obj_id = lock_id;
+			this->log.entries[i].time = g_get_monotonic_time();
+			this->log.entries[i].tier = tier;
+		}
+	}
+
+	entry = (struct KVEntry){
+		.lock_id = lock_id,
+		.backend_id = tier,
+		.generation = 0
+	};
+
+	EXE(kv_put(this, namespace, path, &entry),
+	    "failed to store new object in kv");
+
+	wrapper->scope = (JManagedBackendScope){
+		.mem = wrapper,
+		.lock_id = entry.lock_id,
+		.stack = this,
+		.tier = entry.backend_id,
+		.namespace = namespace,
+		.path = path
+	};
+	*scope = &wrapper->scope;
+
+	ret = TRUE;
+end:
+	return ret;
+}
+
+gboolean
+j_backend_managed_object_close(JManagedBackendScope* this)
+{
+	read_unlock(&g_array_index(this->stack->rw_spin_locks, RWSpinLock, this->lock_id));
+	free(this->mem);
+	return TRUE;
+}
+
+gboolean
+j_backend_managed_policy_message(JManagedBackends* this,
+				 const gchar* type, gpointer data, guint length)
+{
+	if (this->policy->process_message)
+	{
+		return this->policy->process_message(this->policy->data, type, data, length);
+	}
+	return TRUE;
+}
+
+gboolean
+j_backend_managed_get_tiers(JManagedBackends* this, JStorageTier const* const** tiers, guint* length)
+{
+	if (tiers)
+	{
+		*tiers = (JStorageTier const* const*)this->object_tier_data;
+	}
+	*length = this->object_backend_length;
+	return TRUE;
+}
+
+guint
+j_backend_managed_get_tier(JManagedBackends* this,
+			   const gchar* namespace, const gchar* path)
+{
+	struct KVEntry* entry;
+	EXE(kv_get(this, namespace, path, &entry),
+	    "failed to fetch kventry");
+	return entry->backend_id;
+end:
```

Comment:
Other parts of JULEA use `_error` as name for error handling label. Please change to prepare some unified error handling in JULEA.

Reply:
Also in cases below.